### ISSUE TITLE
Incorrect handling of 'finally'?

### ISSUE THREAD

**Issue by GoogleCodeExporter** [typetools/checker-framework#293]
```
Not sure if this is a bug or a misunderstanding on my end.

    public abstract class Init
    {
        public void f()
        {
            String s;
            try {
                s = read();
            }
            catch (Exception e) {
                return;
            }
            finally {
                write("finished");
            }

            write(s);   // <-- error reported here
        }

        public abstract String read() throws Exception;
        public abstract void write(String s);
    }

I expected the nullness checker to report no errors.

# java -jar binary/checkers.jar -processor checkers.nullness.NullnessChecker 
Init.java
Init.java:16: error: incompatible types in argument.
        write(s);
              ^
  found   : @UnknownInitialization @Nullable String
  required: @Initialized @NonNull String
1 error

Removing the 'finally' block entirely makes the error go away.
```

Original issue reported on code.google.com by `kan...@cakoose.com` on 18 Dec 2013 at 1:10

**Comment by mernst** [comment]
I have added another test case to file checker-framework/checker/tests/initialization/fbc/TryFinally.java.  Thanks to Ed Price for the test case.

The problem is that our analysis considers the finally block only once.  On some paths to the finally block, the field has already been initialized, and on other paths to the finally block, the field has not yet been initialized.  Therefore, it conservatively assumes the field has not yet been initialized, resulting in a false positive warning.

As background, each program is represented by an AST (abstract syntax tree) and a control flow graph (CFG).  The dataflow analysis, including initialization analysis is performed over the CFG.  Type-checking is done over the AST (but it should really be done over the CFG).

The problem is that there is only one copy of the finally block in the constructed control flow graph.  The best solution is to duplicate the finally block in the control flow graph.

The challenge with this solution is that the rest of the framework currently assumes that CFG nodes are unique.  Most importantly, when checking the type of an AST expression, we know that it only maps to one CFG node.  If it could map to more than one, then we'd have to do more work to compute the type, perhaps by taking the LUB of all of the copies.  As a more minor point, we would also have to be careful not to produce two different error messages for different CFG nodes that correspond to a single AST expression.

Eventually, we would like to do all type-checking over the CFG, just as the dataflow analysis is currently done over the CFG.  That would make this bug easier to fix.  However, that would be a more substantial refactoring -- it is not the way to get this bug fixed quickly.  The quicker thing to do is to keep typechecking over the AST, but duplicate the finally block in the CFG that is used for dataflow and initialization analysis.

The `-Aflowdotdir` command-line option lets you visualize the CFG, and you can see the single CFG node block for the finally block and the infeasible control flow.

[Note for developers:  there was a conversation "Control flow after finally blocks" from which I gleaned the above information.]

**Comment by stefanheule** [comment]
For all the examples in this thread it seems that it would not be necessary to do all checks on the CFG, and simply analyzing the finally block twice is sufficient. The possible ambiguity for nodes in the finally block should also not be a big issue, as one can conservatively merge all information from the duplicates. One concern might be the exponential blowup in the CFG, though it's only exponential in the nesting level of try-catch-finally statements, which in practice hopefully isn't very high. Also, the desugaring idea discussed here suffers from the same blowup.

To implement this, we could duplicate the finally block in the CFG and properly link the basic blocks so that one is for exceptional behavior and one for normal behavior. This way, we have more precise information after the finally block, because the dataflow analysis knows that the try and finally block both exited normally. An additional change will be that now every AST node may correspond to a set of CFG nodes (instead of a single node), and when asking for dataflow information about an AST node, one would have to merge all the information for all CFG nodes.

I won't be able to implement this feature, but I wanted to provide my perspective in case it is helpful. In particular I wanted to push back a little on the notion that the checking has to be done over the CFG in order to fix this issue. I think that is largely orthogonal (and I have even forgotten why doing the checking over the CFG is advantageous).

**Comment by atomicknight** [comment]
Has there been any further work on this? I'm encountering this issue quite a bit while upgrading from an older version of the framework. Thanks!

**Comment by mernst** [comment]
When this is fixed, please also confirm that the test case in #1291 (which we have closed as a duplicate) now works properly.

**Comment by lazaroclapp** [comment]
So, we are running into some issues with NullAway regarding finally blocks and field initialization, and the best possible fix probably goes right through fixing this issue. I have taken a look at this, the CFG APIs, the earlier fix (#1429) and the issues that it triggered (e.g. #1520).

From what I can tell, the main issue with that fix was the duplication of AST nodes. I feel that a fix which duplicates finally nodes at the CFG level, but not at the AST level, and assigns multiple CFG nodes to the same AST node, should not run into as many issues. This is basically the approach @stefanheule  describes above. 

An issue I can see with that is that `AnalysisResult.getNodeForTree` and the like assume a 1-1 map between AST and CFG nodes, and just changing that might be a very significant API breaking change. On the other hand, it seems at least in theory to be possible to provide both: a "merged" (imprecise, but sound) CFG node for any finally block, and an additional `AnalysisResult.getNodeSetForTree` which retrieves the more precise cloned CFG nodes.

@wmdietl So, before I try implementing this. Do you know/remember if there was a hard reason why the original patch had to duplicate AST nodes other than this 1-1 mapping? Is there anything on the idea above that makes no sense? Any gotchas I am not seeing?

cc: @msridhar

**Comment by msridhar** [comment]
One question I had about the tree cloning approach; does the `treeLookup` API still work for statements inside `finally` blocks?  Or in other words, how does a client get a final, sound dataflow result for statements inside the `finally` blocks?

**Comment by wmdietl** [comment]
For the use within the Checker Framework we might not need this; the AST visitor iterates over each `finally` block separately https://github.com/typetools/checker-framework/blob/master/framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java#L908

I agree that we need to define what analysis results we want to get for statements within a `finally` block.

**Comment by msridhar** [comment]
I see.  Yeah, looking at all `finally` blocks separately works.  We could do something similar in NullAway, but it'll be a bit ugly.  It'd be nice to have the existing Checker dataflow APIs work as usual 😄

**Comment by stefanheule** [comment]
As I wrote above, I think it should work to merge all the information from the different CFG nodes when asking about an AST nodes that has multiple CFG nodes (due to duplication).  This is sound, but less precise than analysing the different blocks separately, and a client could choose whether simplicity or precision is more important for them.
